### ISSUE TITLE
aws lambda installation

### ISSUE THREAD

**Issue by salzhrani** [lovell/sharp#169]
I am trying to use sharp within aws lambda ... I got it running on an EC2 instance and then copied the the `node_modules` with the script as explained [here](https://aws.amazon.com/blogs/compute/nodejs-packages-in-lambda/).. 
when invoking the function I get the following error:
```
module initialization error: Error: libvips.so.40: cannot open shared object file: No such file or directory
 at Module.load (module.js:356:32)
 at Function.Module._load (module.js:312:12)
 at Module.require (module.js:364:17)
 at require (module.js:380:17)
 at Object.<anonymous> (/var/task/node_modules/sharp/index.js:11:13)
 at Module._compile (module.js:456:26)
 at Object.Module._extensions..js (module.js:474:10)
 at Module.load (module.js:356:32)
 at Function.Module._load (module.js:312:12)
 at Module.require (module.js:364:17)
 at require (module.js:380:17)
 at Object.<anonymous> (/var/task/index2.js:4:13)
 at Module._compile (module.js:456:26)
 at Object.Module._extensions..js (module.js:474:10)
 at Module.load (module.js:356:32)
 at Function.Module._load (module.js:312:12)
END RequestId: 4446cc91-bc30-11e4-823c-439adf27cb88
REPORT RequestId: 4446cc91-bc30-11e4-823c-439adf27cb88	Duration: 79.25 ms	Billed Duration: 100 ms 	Memory Size: 128 MB	Max Memory Used: 12 MB	

Message
-------
module initialization error:Error: libvips.so.40: cannot open shared object file: No such file or directory: at Module.load (module.js:356:32)
 at Function.Module._load (module.js:312:12)
 at Module.require (module.js:364:17)
 at require (module.js:380:17)
 at Object.<anonymous> (/var/task/node_modules/sharp/index.js:11:13)
 at Module._compile (module.js:456:26)
 at Object.Module._extensions..js (module.js:474:10)
 at Module.load (module.js:356:32)
 at Function.Module._load (module.js:312:12)
 at Module.require (module.js:364:17)
 at require (module.js:380:17)
 at Object.<anonymous> (/var/task/index2.js:4:13)
 at Module._compile (module.js:456:26)
 at Object.Module._extensions..js (module.js:474:10)
 at Module.load (module.js:356:32)
 at Function.Module._load (module.js:312:12)
```

 Not sure how to get it running properly

**Comment by lovell** [comment]
I've yet to try AWS Lambda, so may not be able to help too much here, but would be interested to learn what's involved in doing so.

It uses Amazon Linux, which according to http://aws.amazon.com/amazon-linux-ami/2014.09-packages/ doesn't include libvips, so I'm guessing you've installed it yourself.

https://aws.amazon.com/blogs/compute/nodejs-packages-in-lambda/ says:

> "You’ll need to either ensure that the libraries and their transitive dependencies are statically compiled"

The `cannot open shared object file` error suggests it's been compiled as a shared library rather than a static library. What flags did you pass to libvips' `./configure` when installing?

> "We specify a prefix of PKG_CONFIG_PATH before we run npm install to include our new static libraries into the local path."

What did you set `PKG_CONFIG_PATH` to?

As an aside, I notice the [Supported Versions](http://docs.aws.amazon.com/lambda/latest/dg/lambda-introduction.html#current-supported-versions) section of the Lamdba docs includes a reference to the unsupported `imagemagick` module on npm.

**Comment by lovell** [comment]
https://twitter.com/salzhrani/status/570235269208809472

**Comment by salzhrani** [comment]
I used the `preinstall.sh` script here to install it ... and it ran fine on EC2.

**Comment by salzhrani** [comment]
I am not sure how to compile it as a static library. I am pretty new to aws.

**Comment by listonb** [comment]
Download libvips and try ./configure --enable-static --disable-shared  and then follow the blog post in regards to this:

PKG_CONFIG_PATH=/location/of/libvips/pkgconfig/  npm install –prefix=~/sharp_demo sharp

Let me know how you fair. You need to either a) compile libvips statically. or b) include the .so files for libvips in the root of your .zip (if they are small enough) and reinstall sharp with: "LDFLAGS="-Wl,-rpath,'/var/task/'" npm install PACKAGENAME)"

Personally I'd recommend the static approach, the rpath is a bit more tricky

**Comment by salzhrani** [comment]
Will try. Thank you.

**Comment by lovell** [comment]
Thanks for helping @listonb and good luck @salzhrani! I'll be interested in finding out more about the steps you had to take when you do get this working.

**Comment by salzhrani** [comment]
I plan to share that once I manage to get it up and running.

**Comment by salzhrani** [comment]
I was able to compile it statically using 
`/configure --enable-debug=no --enable-docs=no --enable-cxx=yes --without-python --without-orc --without-fftw --without-gsf --enable-shared=no --enable-static=yes --prefix=/home/ec2-user/vips`

but when I do an npm install I get the following error
```
[ec2-user@ip-172-31-19-181 fetch]$ PKG_CONFIG_PATH=/home/ec2-user/vips/lib/pkgconfig/ npm install --prefix=/home/ec2-user/fetch/ sharp
\
> sharp@0.9.2 install /home/ec2-user/fetch/node_modules/sharp
> node-gyp rebuild

make: Entering directory `/home/ec2-user/fetch/node_modules/sharp/build'
  CXX(target) Release/obj.target/sharp/src/common.o
  CXX(target) Release/obj.target/sharp/src/utilities.o
  CXX(target) Release/obj.target/sharp/src/metadata.o
  CXX(target) Release/obj.target/sharp/src/resize.o
  CXX(target) Release/obj.target/sharp/src/sharp.o
  SOLINK_MODULE(target) Release/obj.target/sharp.node
/usr/bin/ld: /home/ec2-user/vips/lib/libvips.a(affine.o): relocation R_X86_64_32S against `vips_object_set_property' can not be used when making a shared object; recompile with -fPIC
/home/ec2-user/vips/lib/libvips.a: could not read symbols: Bad value
collect2: error: ld returned 1 exit status
make: *** [Release/obj.target/sharp.node] Error 1
make: Leaving directory `/home/ec2-user/fetch/node_modules/sharp/build'
```

**Comment by salzhrani** [comment]
Also if I try `LDFLAGS="-Wl,-rpath,'/home/ec2-user/fetch'" npm install sharp` I get
```
> sharp@0.9.2 install /home/ec2-user/fetch/node_modules/sharp
> node-gyp rebuild

Package vips was not found in the pkg-config search path.
Perhaps you should add the directory containing `vips.pc'
to the PKG_CONFIG_PATH environment variable
No package 'vips' found
gyp: Call to 'PKG_CONFIG_PATH=":$PKG_CONFIG_PATH:/usr/local/lib/pkgconfig:/usr/lib/pkgconfig" pkg-config --libs vips' returned exit status 1. while trying to load binding.gyp
```

**Comment by lovell** [comment]
The `recompile with -fPIC` message suggests using position-independent code, which you can do via something like `./configure CFLAGS="-fPIC" CXXFLAGS="-fPIC" --enable-debug=no ...`.

**Comment by salzhrani** [comment]
@lovell you are correct ... now there is another error:
```
START RequestId: 819ac5ab-bcf9-11e4-b38b-2b402c743b19
module initialization error: Error: /var/task/node_modules/sharp/build/Release/sharp.node: undefined symbol: WebPEncodeLosslessRGB
 at Module.load (module.js:356:32)
 at Function.Module._load (module.js:312:12)
 at Module.require (module.js:364:17)
 at require (module.js:380:17)
 at Object.<anonymous> (/var/task/node_modules/sharp/index.js:11:13)
 at Module._compile (module.js:456:26)
 at Object.Module._extensions..js (module.js:474:10)
 at Module.load (module.js:356:32)
 at Function.Module._load (module.js:312:12)
 at Module.require (module.js:364:17)
 at require (module.js:380:17)
 at Object.<anonymous> (/var/task/index2.js:4:13)
 at Module._compile (module.js:456:26)
 at Object.Module._extensions..js (module.js:474:10)
 at Module.load (module.js:356:32)
 at Function.Module._load (module.js:312:12)
END RequestId: 819ac5ab-bcf9-11e4-b38b-2b402c743b19
REPORT RequestId: 819ac5ab-bcf9-11e4-b38b-2b402c743b19	Duration: 5.75 ms	Billed Duration: 100 ms 	Memory Size: 128 MB	Max Memory Used: 31 MB	
```

I can sacrifice WebP support if that what is causing it

**Comment by salzhrani** [comment]
the installed package for webp is `Package libwebp-devel-0.3.0-3.5.amzn1.x86_64 already installed and latest version`

**Comment by lovell** [comment]
Are you able to try libwebp >= 0.4? I recall hitting a similar problem trying to process WebP images on a Raspberry Pi (Raspbian).

**Comment by salzhrani** [comment]
I was not able to upgrade webp ... maybe @listonb can help

**Comment by Couto** [comment]
It would be nice if you could publish the steps that you did in order to make this work.
I'm currently trying to make sharp to work on the lambda service, but i'm afraid that my linux skills are not enough to make this work.

**Comment by salzhrani** [comment]
I didn't go any further than the what is mentioned here. Maybe you can take it further.

**Comment by Couto** [comment]
but where you able to make it work or did you just stop trying? just to know if I can try to retrace your steps :)

**Comment by salzhrani** [comment]
I stopped trying. I wasn't able to go any further.

**Comment by lovell** [comment]
If you don't need Webp support, perhaps compile libvips with the additional `./configure --without-webp ...` flag to avoid the `undefined symbol: WebPEncodeLosslessRGB` error.

**Comment by bozzltron** [comment]
I'm trying to deploy a node application that uses the sharp module to heroku via docker.  This thread has been extremely helpful.  However, I haven't resolved the WebPEncodeLosslessRGB error.  I've added the flag that you have suggested.  Please take a look at my Dockerfile.   Thanks in advance!

https://github.com/bozzltron/koa-sharp/blob/master/Dockerfile

**Comment by lovell** [comment]
@bozzltron This discussion relates to the specifics of the AWS Lambda environment, which requires the use of static linking.

You should be able to get away with dynamic linking on Heroku's "cedar" stack. Have you tried https://registry.hub.docker.com/u/marcbachmann/libvips/dockerfile/ ?

Feel free to open a new issue to discuss this further.

**Comment by bozzltron** [comment]
Dynamic linking fails at runtime on heroku just like in this aws thread.  I have tried this file in the past.  I can't copy it directly as its not cedar based.  Webp is only thing standing in my way.

**Comment by lovell** [comment]
@bozzltron With Docker inside Heroku inside LXC (inside ...), it's [containers all the way down](https://en.wikipedia.org/wiki/Turtles_all_the_way_down) :turtle: 

Do you need WebP support?

**Comment by bozzltron** [comment]
That's makes sense.  You're container doesn't have to be cedar based!  You can put whatever you like in there.  Unfortunately, I tried releasing my app based on the docker image you suggested and its has the same "cannot open shared object file" problem that I had already solved.

I do not need webp and I've followed your suggestion to compile with out it via the flag.  But that hasn't fixed the issue.

**Comment by lovell** [comment]
Quick thought: try removing [libwebp-dev](https://github.com/bozzltron/koa-sharp/blob/master/Dockerfile#L7) from the list of `apt-get install`ed packages.

**Comment by bozzltron** [comment]
I made the change and tried the release.  At run time, I now get this "undefined symbol: jpeg_resync_to_restart"  I hope this is not a rabbit hole of undefined symbols!  Thanks for all your help.

**Comment by acuervof** [comment]
Before attempting to implement a lambda function with sharp, could anyone use the library in aws lambda?? I haven't seen any github project. Thanks!

**Comment by vlapo** [comment]
Hi @bozzltron, 
I had sam problem on runtime. As workaround you have to edit vips.pc file before npm install.
Change vips.pc in pkgconfig folder from:
```
prefix=/tmp/compile/vips10
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: vips
Description: Image processing library
Version: 8.1.0
Requires.private:  glib-2.0 libxml-2.0 gmodule-2.0 gobject-2.0 libtiff-4 libpng
Requires: gobject-2.0
Libs: -L${libdir} -lvips
Libs.private:  -ljpeg
Cflags: -I${includedir}
```
To:
```
prefix=/tmp/compile/vips10
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: vips
Description: Image processing library
Version: 8.1.0
Requires.private:  glib-2.0 libxml-2.0 gmodule-2.0 gobject-2.0 libtiff-4 libpng
Requires: gobject-2.0 glib-2.0 libxml-2.0 gmodule-2.0 gobject-2.0 libtiff-4 libpng
Libs: -L${libdir} -lvips -ljpeg
Libs.private:  -ljpeg
Cflags: -I${includedir}
```
and after edit run npm install command:
```PKG_CONFIG_PATH=/path_to_static_compiled_vips/lib/pkgconfig/ npm install --prefix=/sharp sharp```
There are no runtime errors after this steps.

I think, it`s because libvips use Libs.private and Requires.private properties in pc file (because https://github.com/jcupitt/libvips/issues/196) and sharp has to compile with --static arg in pkg-config command.

**Comment by e-tip** [comment]
@vlapo i've modified vips.pc but now i get this error
"errorMessage": "libjpeg.so.8: cannot open shared object file: No such file or directory", when i try with vips 8.1.1 version does anyone have a suggestion ?

**Comment by vlapo** [comment]
Did you install all libraries ? 
```
sudo yum install -y gtk-doc libxml2-devel libjpeg-turbo-devel libpng-devel libtiff-devel libexif-devel libgsf-devel lcms2-devel ImageMagick-devel gobject-introspection-devel libwebp-devel curl
```

**Comment by e-tip** [comment]
Yes all dependencies are installed. May it depend because i'm compiling against ubuntu libraries ?

**Comment by lovell** [comment]
It looks like @acuervof has had some success with the new `look` work-in-progress branch. It installs libvips and its dependencies as shared libraries inside `node_modules` - see https://github.com/lovell/sharp/issues/42#issuecomment-152134367

**Comment by felixbuenemann** [comment]
This page list the [AMIs for Lambda](http://docs.aws.amazon.com/lambda/latest/dg/current-supported-versions.html), so I suggest setting up a small EC2 instance with the same AMI to compile any native code.

**Comment by felixbuenemann** [comment]
I had no trouble to run sharp on lamba. I installed the packages on an EC2 instance with a matching Amazon Linux AMI and the node.js 0.10.36 from the epel repo paired with the latest npm 3.7.3 via their install.sh script. I had no luck with the ancient npm 1.4.x from the epel repo as it failed to compile the nan npm dependency.

So I think this issue can be closed.

**Comment by tc** [comment]
@felixbuenemann do you have the sample project of sharp on lambda on github?

**Comment by felixbuenemann** [comment]
The current version has way too much stuff, like temp files to handle tiff, download and upload md5 verification etc.

But I have a simple early prototype that uses in memory buffers and is very simple to read. It's adapted from the AWS [Using AWS Lambda with Amazon S3](http://docs.aws.amazon.com/lambda/latest/dg/with-s3-example.html) Tutorial, which uses the `gm` npm package (and is a bazillion times slower and more memory hungry than sharp):

I'm a rubyist with barely any node.js knowledge, so the code is less pretty than it could be ;-)

```javascript
// dependencies
var async = require('async');
var AWS = require('aws-sdk');
var sharp = require('sharp');
var util = require('util');

// constants
var MAX_WIDTH  = 256;
var MAX_HEIGHT = 256;
var STORAGE_CLASS = 'REDUCED_REDUNDANCY';

// get reference to S3 client 
var s3 = new AWS.S3();

var mimeTypes = {
    "jpg": "jpeg",
    "jpe": "jpeg",
    "jpeg": "jpeg",
    "png": "png",
    "gif": "gif"
};

exports.handler = function(event, context) {
    // Read options from the event.
    console.log("Reading options from event:\n", util.inspect(event, {depth: 5}));
    var srcBucket = event.Records[0].s3.bucket.name;
    // Object key may have spaces or unicode non-ASCII characters.
    var srcKey    = decodeURIComponent(event.Records[0].s3.object.key.replace(/\+/g, " "));
    var dstBucket = srcBucket.replace(/-source$/, '-target');
    var dstPrefix = 'thumb-';
    var dstKey    = srcKey.replace(/(\/|^)([^/]+$)/, '$1' + dstPrefix + '$2');

    // Sanity check: validate that source and destination are different buckets.
    if (srcBucket == dstBucket) {
        console.error("Destination bucket must not match source bucket.");
        context.succeed(); return;
    }

    // Infer the image type.
    var typeMatch = srcKey.match(/\.([^.]*)$/);
    if (!typeMatch) {
        console.error('unable to infer image type for key ' + srcKey);
        context.succeed(); return;
    }
    var imageType = typeMatch[1];
    if (!/^(jpg|png|gif|webp)$/.test(imageType)) {
        console.log('skipping unsupported file ' + srcKey);
        context.succeed(); return;
    }
    var format = mimeTypes[imageType];
    if (!format) {
        console.error('missing mime type mapping for extension ' + imageType);
        context.succeed(); return;
    }

    // Download the image from S3, transform, and upload to a different S3 bucket.
    async.waterfall([
        function download(next) {
            // Download the image from S3 into a buffer.
            s3.getObject({
                Bucket: srcBucket,
                Key: srcKey
            },
            next);
        },
        function transform(response, next) {
            sharp(response.Body)
                .resize(MAX_WIDTH, MAX_HEIGHT).max()
                .toFormat(format)
                .toBuffer(next);
        },
        function upload(data, info, next) {
            var contentType = 'image/' + format;
            // Stream the transformed image to a different S3 bucket.
            s3.putObject({
                Bucket: dstBucket,
                Key: dstKey,
                Body: data,
                ContentType: contentType,
                StorageClass: STORAGE_CLASS
            },
            next);
        }
        ], function (err, result) {
            // Stripped some logging here...
            context.done(err, result);
        }
    );
};
```
I've uploaded the full deployment package [CreateThumbnailVipsBuffer.zip](https://www.dropbox.com/s/2kit09n5y4p447n/CreateThumbnailVipsBuffer.zip?dl=0) including the package.json which is only needed locally. Be sure to actually check the content before blindly uploading it to lambda…

**Comment by lovell** [comment]
@felixbuenemann @tc If you're able to share details publicly, it's always great to learn more about what people are using `sharp` for over at #35

**Comment by tc** [comment]
thanks @felixbuenemann !

@lovell will do once i ship it :)

**Comment by lovell** [comment]
Die erstaunliche @felixbuenemann has very kindly provided [Lambda install docs](http://sharp.dimens.io/en/latest/install/#aws-lambda).
